### PR TITLE
The Raulnarok PR of 2018

### PR DESCRIPTION
**Pull request by raulverag**
This PR does a number of things, which couldn't really be broken out into smaller pieces:
- It moves all of runtime, platform, and tracelib under a src top-level directory
- It creates a src/planning directory and moves strategizer.ts into it
- It modifies the TS compiler config to take input from src/runtime and src/planning, and put it's output in parallel directories under build
- It makes imports of non-TS js files by TS files work by hard-linking a long list of stuff from src to build, so that imports during compile find the files in src, and imports at runtime find the same files in build.

All tests pass. 

There is much more to do. Most notably, the linking mechanism should go away as soon as possible. As JS is converted to TS, entries for the JS files should be removed from the list. Hence the name of the list file is 'tools/reducethislist'. Many files are linked that aren't js, (mostly test artifacts) because they live in the src tree and are loaded at runtime. These should be moved out, or the loads adjusted, so links in the build tree aren't necessary. There is lots of low-hanging fruit. The big win is that there are no longer any ts or ts-build directories, and ts files can live anywhere, as it's now very easy to add a new directory to the tsc input.

### CODE REVIEW

File: shell/apps/remote-planning/runtime/mock-slot-composer.js [hunk 1]
```diff
@@ -9,8 +9,11 @@
  */
 'use strict';
 
-import {SlotComposer} from '../../../../runtime/ts-build/slot-composer.js';
-import {SlotDomConsumer} from '../../../../runtime/ts-build/slot-dom-consumer.js';
+const arcs = '../../../..';
+const runtime = '${arcs}/build/runtime';
+
+import {SlotComposer} from '${runtime}/slot-composer.js';
+import {SlotDomConsumer} from '${runtime}/slot-dom-consumer.js';
```
Comment (mykmartin) [inline shell/apps/remote-planning/runtime/mock-slot-composer.js]: The 2 imports and the const runtime above need to be back-ticked.

File: shell/apps/remote-planning/shell/browser-loader.js [hunk 2]
```diff
@@ -8,11 +8,14 @@
...
+
 import {Loader} from '../runtime/loader.js';
-import {Particle} from '../../../../runtime/ts-build/particle.js';
-import {DomParticle} from '../../../../runtime/dom-particle.js';
-import {MultiplexerDomParticle} from '../../../../runtime/multiplexer-dom-particle.js';
-import {TransformationDomParticle} from '../../../../runtime/transformation-dom-particle.js';
+import {Particle} from '${runtime}/particle.js';
+import {DomParticle} from '${runtime}/dom-particle.js';
+import {MultiplexerDomParticle} from '${runtime}/multiplexer-dom-particle.js';
+import {TransformationDomParticle} from '${runtime}/transformation-dom-particle.js';
```
Comment (mykmartin) [inline shell/apps/remote-planning/shell/browser-loader.js]: Back-ticks also.

File: shell/source/worker-entry.js [hunk 3]
```diff
@@ -6,7 +6,7 @@
 // subject to an additional IP rights grant found at
 // http://polymer.github.io/PATENTS.txt
 
-import {ParticleExecutionContext} from '../../runtime/ts-build/particle-execution-context.js';
+import {ParticleExecutionContext} from '../..//build/runtime/particle-execution-context.js';
```
Comment (mykmartin) [inline shell/source/worker-entry.js]: Double slash: ..//build

File: tools/spotify-importer.js [hunk 4]
```diff
@@ -8,11 +8,13 @@
...
+ 
 import fs from 'fs';
-import {Manifest} from '../runtime/manifest.js';
-import {Type} from '../runtime/ts-build/type.js';
-import {StorageProviderFactory} from '../runtime/ts-build/storage/storage-provider-factory.js';
-import {resetStorageForTesting} from '../runtime/ts-build/storage/firebase-storage.js';
+import {Manifest} from '${runtime}/manifest.js';
+import {Type} from '.${runtime}/type.js';
+import {StorageProviderFactory} from '${runtime}/storage/storage-provider-factory.js';
+import {resetStorageForTesting} from '${runtime}/storage/firebase-storage.js';
```
Comment (mykmartin) [inline tools/spotify-importer.js]: More backticks (although I would prefer not to use a global var at all and just inline the paths directly).

File: shell/apps/remote-planning/runtime/mock-slot-composer.js (hunk 1 above)
Comment (raulverag) [reply shell/apps/remote-planning/runtime/mock-slot-composer.js]: Travis found this. Fixed.

File: shell/apps/remote-planning/shell/browser-loader.js (hunk 2 above)
Comment (raulverag) [reply shell/apps/remote-planning/shell/browser-loader.js]: Ditto.

File: shell/source/worker-entry.js (hunk 3 above)
Comment (raulverag) [reply shell/source/worker-entry.js]: Thanks

File: tools/spotify-importer.js (hunk 4 above)
Comment (raulverag) [reply tools/spotify-importer.js]: Done.

**Comment by mykmartin** [comment]
Nice work Raul!

**Pull request merged**